kad-dht/README.md

```diff
@@ -0,0 +1,395 @@
+Kademlia paper [0]. Peer IDs are normalised through the SHA256 hash function.
+
+For recap, `dXOR(sha256(id1), sha256(id2))` is the number of common leftmost
+bits between SHA256 of each peer IDs. The `dXOR` between us and a peer X
+designates the bucket index that peer X will take up in the Kademlia routing
+table.
+
+## Kademlia routing table
+
+The data structure backing this system is a k-bucket routing table, closely
+following the design outlined in the Kademlia paper [0]. The default value for
+`k` is 20, and the maximum bucket count matches the size of the SHA256 function,
+i.e. 256 buckets.
+
+The routing table is unfolded lazily, starting with a single bucket a position 0
+(representing the most distant peers), and splitting it subsequently as closer
+peers are found, and the capacity of the nearmost bucket is exceeded.
+
+## Alpha concurrency factor (α)
+
+The concurrency of node and value lookups are limited by parameter `α`, with a
+default value of 3. This implies that each lookup process can perform no more
+than 3 inflight requests, at any given time.
+
+## Record keys
+
+Records in the DHT are keyed by CID [4], roughly speaking. There are intentions
+to move to multihash [5] keys in the future, as certain CID components like the
+multicodec are redundant. This will be an incompatible change.
+
+The format of `key` varies depending on message type; however, in all cases
+`dXOR(sha256(key1), sha256(key2))` see [Distance function](#distance-function-dxor)
+is used as the distance between two keys.
+
+* For `GET_VALUE` and `PUT_VALUE`, `key` is an unstructured array of bytes, except
+if it is being used to look up a public key for a `PeerId`, in which case it is
+the ASCII string '/pk/' concatenated with the binary `PeerId`.
+* For `ADD_PROVIDER` and `GET_PROVIDERS`, `key` is interpreted and validated as
+a CID.
+* For `FIND_NODE`, `key` is a binary `PeerId`
+
+## Interfaces
+
+The libp2p Kad DHT implementation satisfies the routing interfaces:
+
+```go
+type Routing interface {
+	ContentRouting
+	PeerRouting
+	ValueStore
+
+	// Kicks off the bootstrap process.
+	Bootstrap(context.Context) error
+}
+
+// ContentRouting is used to find information about who has what content.
+type ContentRouting interface {
+	// Provide adds the given CID to the content routing system. If 'true' is
+	// passed, it also announces it, otherwise it is just kept in the local
+	// accounting of which objects are being provided.
+	Provide(context.Context, cid.Cid, bool) error
+
+	// Search for peers who are able to provide a given key.
+	FindProvidersAsync(context.Context, cid.Cid, int) <-chan pstore.PeerInfo
+}
+
+// PeerRouting is a way to find information about certain peers.
+//
+// This can be implemented by a simple lookup table, a tracking server,
+// or even a DHT (like herein).
+type PeerRouting interface {
+	// FindPeer searches for a peer with given ID, returns a pstore.PeerInfo
+	// with relevant addresses.
```

Comment:
Why is addresses plural?

Reply:
Because peers can have multiple addresses: different ports, protocols, locations etc.